add wet day frequency correction

dgergel opened the pull request:
This PR adds a wet day frequency correction service. 

The wet day frequency method follows the Cannon et al 2015 method: motivated by the mixed discrete continuous nature of precip data, dry days are treated as values below a set threshold (in that paper, 0.05 mm/day). Zero values in the observed and modeled data are replaced with nonzero uniform random values below the threshold before bias correction; after bias correction (and downscaling in our case), all values below the threshold are set to zero. The pre and post bias correction options are denoted here with `pre-process` and `post-process` CLI options. 

Basic CLI interface is: 

```
dodola correct-wetday-frequency /path/to/store.zarr \
    --out /path/to/corrected.zarr \
    --process "pre-process" / "post-process"
```

closes #40

kemccusker commented on dodola/tests/test_services.py:
```diff
@@ -351,6 +352,47 @@ def test_remove_leapdays():
+
+    if process == "pre-process":
+        # all 0 values should have been set to a random uniform value below 0.05
+        assert (
+            ds_precip_corrected["fakevariable"].where(
+                ds_precip["fakevariable"] == 0, drop=True
+            )
+            != 0.0
+        )
```
Should these `assert`s have `.all()` ? Same for lines L380-393
```suggestion
        assert ((
            ds_precip_corrected["fakevariable"].where(
                ds_precip["fakevariable"] == 0, drop=True
            )
            != 0.0).all()
        )
```

dgergel replied:
Because the test time series only has one 0, that's why I didn't include the `.all()`s here

brews commented on dodola/cli.py:
```diff
@@ -191,3 +191,18 @@ def regrid(x, out, method, domain_file, weightspath):
+    "--process",
+    "-p",
+    required=True,
+    type=click.Choice(["pre-process", "post-process"], case_sensitive=False),
```
Minor suggestion to shorten things because you already have "--process" as part of the input.

```suggestion
    type=click.Choice(["pre", "post"], case_sensitive=False),
```

dgergel replied:
Great idea, thanks!

brews replied:
@dgergel Thanks but your other functions in core, services, and tests are still expecting (and documented for) `"pre-process"` and `"post-process"`. I think you'd need to switch those over for this to work.

brews commented on dodola/core.py:
```diff
@@ -231,3 +232,26 @@ def xclim_remove_leapdays(ds):
+    Parameters
+    ----------
+    ds : xr.Dataset
+    process : str
```
Document the choices available:

```suggestion
    process : {"pre-process", "post-process"}
```

brews commented on dodola/services.py:
```diff
@@ -244,6 +245,24 @@ def remove_leapdays(x, out):
+    out : str
+        Storage URL to write regridded output to.
+    process : {"pre-process", "post-process"}
+        Step in pipeline, used in determining how to correct.
```
Either/both in `dodola/services.py` or `dodola/core.py`, it would be nice to elaborate a  little bit more on what this 'process' arg does or when it's used. What's the how and why? As a user I'm thinking "is this a clean step thing (it's a 'correction')? or when do each of these pre/post steps occur?"

dgergel replied:
Thanks for this, adding to `dodola/services.py`.

brews commented on dodola/tests/test_services.py:
```diff
@@ -351,6 +352,47 @@ def test_remove_leapdays():
+    ds_precip_corrected = repository.read(out_url)
+
+    if process == "pre-process":
+        # all 0 values should have been set to a random uniform value below 0.05
```
Good comment. Thank you!

brews commented on dodola/cli.py:
```diff
@@ -191,3 +191,18 @@ def regrid(x, out, method, domain_file, weightspath):
+)
+def correct_wetday_frequency(x, out, process):
+    """Correct wet day frequency in a dataset"""
+    services.correct_wet_day_frequency(str(x), out=str(out), process=str(process))
```
Potentially bigger issue with implications across files:

Do we need get a user input variable name, and extract/correct a DataArray for that variable, returning a Dataset with the corrected variable?

As it is right now, couldn't this correction apply to everything in the input Dataset? (and that's potentially a disaster...?)

dgergel replied:
Right now it is up to the user to only apply this to precip. Do we want to raise an error if the variable in a dataset is _not_ precip?

brews replied:
Ah, @dgergel , first I was thinking that there could be an unexpected side effect if this filtered a coordinate variable (lat, lon, height) ...but I couldn't reproduce it with a reasonable bug in a quick test, so I'll drop it. That's bogus.

The other case where this might be a problem is if someone kept temp and precip in the same zarr store. The user would be able to say "yo, only do this for the `precip` variable" -- so that it wouldn't filter the temp data available in the same store.

This is a bit of a reach so I wouldn't change it unless you see the need.

brews reviewed (changes requested):
I think this looks pretty good!

I have a few inline comments. Some about things that might break stuff. In addition:

- [ ] Could you add an additional test-case to `dodola/tests/test_cli.py`?  In the `test_cli_helpflags()` parametrize decorator could you add `"correct-wetday-frequency"` to the first list and `"correct-wetday-frequency --help"` to the tuple of ids. This will run the CLI and ensure that the subcommand spits out help and not an error.
- [ ] Could you add an entry for this PR to HISTORY.rst?

Finally, should we cite the paper we're using for this method in the `dodola/core.py` function docstr? A la https://numpydoc.readthedocs.io/en/latest/format.html#references. We've been pretty bad about that. I was thinking about it here because we have a reference in the PR description.

dgergel commented on dodola/tests/test_services.py:
```diff
@@ -351,6 +352,47 @@ def test_remove_leapdays():
+    n = 1500
+    threshold = 0.05
+    ts = np.linspace(0.0, 10, num=n)
+    ds_precip = _datafactory(ts, start_time="1950-01-01")
```
Definitely true, this was an artifact from other tests.

dgergel commented:
@brews and @kemccusker, I've addressed your review comments! The one outstanding question is if we want to raise an error if the user attempts to correct something that isn't precip. Right now it is up to the user to only apply this correction to a precip variable but we could also just add an error message in to check this.

brews commented on dodola/services.py:
```diff
@@ -318,6 +319,24 @@ def remove_leapdays(x, out):
+    out : str
+        Storage URL to write regridded output to.
+    process : {"pre-process", "post-process"}
+        Step in pipeline, used in determining how to correct. If pre-process, replaces all zero values with a uniform random value below a threshold. If post-process, replaces all values below a threshold with zeroes. Pre-process occurs before bias correction and downscaling, post-process occurs after.
```
Minor suggestion. Wrap this line at 80 characters?

brews commented on dodola/core.py:
```diff
@@ -319,3 +321,30 @@ def xclim_remove_leapdays(ds):
+    Parameters
+    ----------
+    ds : xr.Dataset
+    process : {"pre-process", "post-process"}
```
```suggestion
    process : {"pre", "post"}
```

dgergel replied:
Good catch, thanks!

brews commented on dodola/core.py:
```diff
@@ -319,3 +321,30 @@ def xclim_remove_leapdays(ds):
+    """
+    threshold = 0.05  # mm/day
+    low = 1e-16
+    if process == "pre-process":
```
```suggestion
    if process == "pre":
```

brews commented on dodola/core.py:
```diff
@@ -319,3 +321,30 @@ def xclim_remove_leapdays(ds):
+    low = 1e-16
+    if process == "pre-process":
+        ds_corrected = ds.where(ds != 0.0, np.random.uniform(low=low, high=threshold))
+    elif process == "post-process":
```
```suggestion
    elif process == "post":
```

brews commented on dodola/services.py:
```diff
@@ -318,6 +319,24 @@ def remove_leapdays(x, out):
+        Storage URL to input xr.Dataset that will be regridded.
+    out : str
+        Storage URL to write regridded output to.
+    process : {"pre-process", "post-process"}
```
```suggestion
    process : {"pre", "post"}
```

brews commented on dodola/tests/test_services.py:
```diff
@@ -441,6 +442,47 @@ def test_remove_leapdays():
 
 
+@pytest.mark.parametrize(
+    "process", [pytest.param("pre-process"), pytest.param("post-process")]
```
```suggestion
    "process", [pytest.param("pre"), pytest.param("post")]
```

brews commented on dodola/tests/test_services.py:
```diff
@@ -441,6 +442,47 @@ def test_remove_leapdays():
+    correct_wet_day_frequency(in_url, out=out_url, process=process)
+    ds_precip_corrected = repository.read(out_url)
+
+    if process == "pre-process":
```
```suggestion
    if process == "pre":
```

brews commented on dodola/tests/test_services.py:
```diff
@@ -441,6 +442,47 @@ def test_remove_leapdays():
+            )
+            < threshold
+        )
+    elif process == "post-process":
```
```suggestion
    elif process == "post":
```

brews reviewed (changes requested):
Needs some very minor changes:

One bug: I marked a few spots where "pre" and "post" are needed for the "processing" arg. The CLI args won't work without these changes.

Also there is one spot where the docstr runs on and on and on and on. Aside from that this looks good to me!

dgergel commented:
@brews I think this should be good to go!

brews reviewed (approved):
I think this is good to go, @dgergel! Appreciate the update. Merge whenever you're ready.
